lib/Dialect/TritonGPU/Transforms/Pipeliner/WGMMAPipeline.cpp

```diff
@@ -206,6 +233,172 @@ static void threadValuesThroughWait(ttng::WarpGroupDotWaitOp wait,
+  llvm::append_range(transOrder, llvm::reverse(llvm::seq(
+                                     rank - 1, (int64_t)shape.size() - 1)));
+  lhs = builder.create<tt::TransOp>(loc, lhs, transOrder);
+  // We split recursively
+  SmallVector<Value> curr;
+  SmallVector<Value> ret = {lhs};
+  for (int i = 1; i < nSplits; i *= 2) {
+    curr = ret;
+    ret.clear();
+    for (auto v : curr) {
+      auto split = builder.create<tt::SplitOp>(loc, v);
+      ret.push_back(split.getResult(0));
+      ret.push_back(split.getResult(1));
+    }
+  }
+
+  auto mmav3Type =
+      type.clone(cast<RankedTensorType>(ret.front().getType()).getShape());
+  // Convert the LHS to mmav3 layout
+  for (auto &v : ret) {
+    v = builder.create<ttg::ConvertLayoutOp>(loc, mmav3Type, v);
+    // These convert_layout ops are noops by construction
+    assert(isNoop(v.getDefiningOp()));
+  }
+  assert(ret.size() == nSplits);
+  return ret;
+}
+
+// Split the RHS of a RSWGMMADot operation into multiple multiple
+// tensors of size newKxN via MemDescSubview
+SmallVector<Value> splitRhs(OpBuilder &builder,
+                            TypedValue<ttg::MemDescType> rhs, int64_t newK) {
+  auto loc = rhs.getLoc();
+  auto *ctx = builder.getContext();
+  auto type = rhs.getType();
+  auto rank = type.getRank();
+  auto kDim = rank - 2;
+  auto nSplits = type.getShape()[kDim] / newK;
+  // offset -> matrix
+  auto ll = ttg::toLinearLayout(type.getShape(), type.getEncoding());
+  auto llInv = ll.invert();
+
+  // Split into
+  auto kOffset = StringAttr::get(ctx, "offset");
+  assert(llInv.getOutDimSize(kOffset) == product(type.getShape()));
+  auto dimNames = tt::standardOutDimNames(ctx, rank);
+  SmallVector<std::pair<StringAttr, int32_t>> newOutDims;
+  for (auto d : getOrder(type)) {
+    newOutDims.push_back({dimNames[d], type.getShape()[d]});
+  }
+  // Split into shmem shape and invert
+  llInv = llInv.reshapeOuts(newOutDims);
+  llInv = llInv.transposeOuts(dimNames);
+  auto toOffsets =
+      [&](const SmallVector<std::pair<StringAttr, int32_t>> &shape) {
+        return llvm::to_vector(llvm::map_range(shape, [&](const auto &p) {
+          return builder.create<arith::ConstantIntOp>(loc, p.second, 32)
+              .getResult();
+        }));
+      };
+  // New Shape
+  auto shape = llvm::to_vector(type.getShape());
+  shape[kDim] = newK;
+  auto newType = ttg::MemDescType::get(
+      shape, type.getElementType(), type.getEncoding(), type.getMemorySpace(),
+      /*isMutable=*/false, type.getAllocShape());
+  SmallVector<Value> ret;
+  SmallVector<std::pair<StringAttr, int32_t>> logicalOffsets;
+  for (int i = 0; i < rank; i++) {
+    logicalOffsets.push_back({StringAttr::get(ctx, "dim" + Twine(i)), 0});
+  }
+  for (int i = 0; i < nSplits; i++) {
+    logicalOffsets[kDim].second = i * newK;
+    auto shmemOffsets = toOffsets(llInv.apply(logicalOffsets));
+
+    Value newSmem = builder.create<triton::gpu::MemDescSubviewOp>(
+        loc, newType, rhs, shmemOffsets);
```

Comment:
Can you show me an example IR after this branch?

My understanding is that now the sliced dims are not as intuitive as they were before; we directly encode the physical memory offsets in mem desc subview operations.

Reply:
Yeah. The longer-term consideration is that if we are going to propose the lower-level DSL, how do we expose mem subview operators?

Reply:
100%. We need to make this abstraction more robust. Probably a better take would be to restrict subviews to contiguous subviews. Either that or deploy some LL-based system of strides for subviews. We can discuss this once we start fleshing out the DSL.

Reply:
Talked to Keren and I don't think we want to change the semantic of subview to take physical offset, it feels like wrong kind of layering as changing the layout would change the offsets.
it also makes the meaning of the IR very non-intuitive and doesn't match how the rest of the IR works. 

One way to support things this way would be to keep track of the multi-dim offsets relative to the beginning of the alloc but that might be a big change.
In the meantime could we verify that we don't support  slicing within a swizzling pattern?

Reply:
Ah, so the semantics are wrt. the original tensor then? That makes sense. It would basically account to bringing the logic I wrote in the pipeliner to here.
Let me change that and tighten the verifier in the meantime.

Reply:
yes, sorry, I thought this was commenting the view lowering. Will have a PR giving proper semantics to this function soon.

Reply:
thanks :)